File: stats.go
package groundcover

import "github.com/groundcover-com/groundcover-go/internal/selfmetrics"

// Stats is a point-in-time snapshot of the SDK's self-observability counters.
// Field names mirror the exported Prometheus metric names.
type Stats struct {
	// Captured is the number of events accepted into the pipeline.
	Captured int64
	// Sent is the number of events successfully delivered.
	Sent int64
	// DroppedOverflow is the number of events evicted by buffer overflow.
	DroppedOverflow int64
	// DroppedSendExhausted is the number of events dropped after delivery failed.
	DroppedSendExhausted int64
	// DroppedBeforeSend is the number of events dropped by BeforeSend.
	DroppedBeforeSend int64
	// Retries is the number of delivery retry attempts.
	Retries int64
	// RateLimited is the number of 429 responses observed.
	RateLimited int64
	// PanicsRecovered is the number of recovered SDK-internal panics.
	PanicsRecovered int64
	// ConfigReloads is the number of configuration swaps.
	ConfigReloads int64
	// QueuePendingItems is the current number of buffered events.
	QueuePendingItems int64
	// QueuePendingBytes is the current estimated size of buffered events.
	QueuePendingBytes int64
	// SubsystemsDisabled is the number of background subsystems self-disabled
	// after a panic.
	SubsystemsDisabled int64
}

// statsFromSnapshot maps an internal counter snapshot to the public Stats type.
func statsFromSnapshot(s selfmetrics.Snapshot) Stats {
	return Stats{
		Captured:             s.Captured,
		Sent:                 s.Sent,
		DroppedOverflow:      s.DroppedOverflow,
		DroppedSendExhausted: s.DroppedSend,
		DroppedBeforeSend:    s.DroppedBeforeSend,
		Retries:              s.Retries,
		RateLimited:          s.RateLimited,
		PanicsRecovered:      s.PanicsRecovered,
		ConfigReloads:        s.ConfigReloads,
		QueuePendingItems:    s.QueuePendingItems,
		QueuePendingBytes:    s.QueuePendingBytes,
		SubsystemsDisabled:   s.SubsystemsDisabled,
	}
}

File: groundcover-python/src/groundcover/_stats.py
"""Public self-observability counters."""

from __future__ import annotations

import dataclasses

from ._internal.selfmetrics import Snapshot


@dataclasses.dataclass(frozen=True)
class Stats:
    """A point-in-time snapshot of the SDK's self-observability counters.
    Field names mirror the exported Prometheus metric names."""

    captured: int = 0
    """The number of events accepted into the pipeline."""
    sent: int = 0
    """The number of events successfully delivered."""
    dropped_overflow: int = 0
    """The number of events evicted by buffer overflow."""
    dropped_send_exhausted: int = 0
    """The number of events dropped after delivery failed."""
    dropped_before_send: int = 0
    """The number of events dropped by before_send."""
    retries: int = 0
    """The number of delivery retry attempts."""
    rate_limited: int = 0
    """The number of 429 responses observed."""
    panics_recovered: int = 0
    """The number of contained SDK-internal faults."""
    config_reloads: int = 0
    """The number of configuration swaps."""
    queue_pending_items: int = 0
    """The current number of buffered events."""
    queue_pending_bytes: int = 0
    """The current estimated size of buffered events."""
    subsystems_disabled: int = 0
    """The number of background subsystems self-disabled after a fault."""


def stats_from_snapshot(s: Snapshot) -> Stats:
    """Map an internal counter snapshot to the public Stats type."""
    return Stats(
        captured=s.captured,
        sent=s.sent,
        dropped_overflow=s.dropped_overflow,
        dropped_send_exhausted=s.dropped_send,
        dropped_before_send=s.dropped_before_send,
        retries=s.retries,
        rate_limited=s.rate_limited,
        panics_recovered=s.panics_recovered,
        config_reloads=s.config_reloads,
        queue_pending_items=s.queue_pending_items,
        queue_pending_bytes=s.queue_pending_bytes,
        subsystems_disabled=s.subsystems_disabled,
    )
